File: Java/MergeSort.java
public class MergeSort {
    /*
     * run - time complexity - O(n log n ) 
     * space complexity - O(n)
     * this will use more memory 
     * 
     */

    public static void main(String[] args) {
        int numbers[] = {9,8,7,6,5,4,3,2,1};
        mergeSort(numbers); // calling merge sort
        for (int i : numbers){
            System.out.print(i+" ");
        }        
    }

    private  static  void mergeSort(int[] array){
        int length = array.length;
        if (length <= 1 )
            return; // base case 
        int middle = length / 2;
        int leftArray[] = new int[middle];
        int rightArray[] = new int[length - middle];
        int i = 0; // left array
        int j = 0; // right array
        // deviding array to right and left arrays
        for (;i  < length; i++ ){
            if (i < middle){
                leftArray[i]=array[i];
            }
            else{
                rightArray[j] = array[i];
                j++;
            }
        }
        mergeSort(leftArray);
        mergeSort(rightArray); // recrusionly calling merge method

        merge(leftArray, rightArray, array);
    }
    private  static  void merge(int[] leftArray,int[] rightArray,int[] array){
        int leftSize = array.length / 2;
        int rightSize = array.length - leftSize;
        int i = 0, l = 0, r = 0; // for original array,left right

        // check the conditions for merging
        while (l < leftSize && r < rightSize) {
            // comparing elements 
            if (leftArray[l] < rightArray[r]){
                array[i] = leftArray[l]; 
                i++;
                l++;
            } else {
                array[i] = rightArray[r];
                i++;
                r++;
            }
        }
        
        while (l < leftSize) { 
            array[i] = leftArray[l];
            i++;
            l++;
        }
        while (r < rightSize){
            array[i] = rightArray[r];
            i++;
            r++;
        }  

    }
    
}

File: simple C/mergeSort.c
#include <stdio.h>
#include <stdlib.h>
// merge arrays
void merge(int *left, int leftSize, int *right, int rightSize, int *array) {
    int i = 0, l = 0, r = 0;
    while (l < leftSize && r < rightSize) {
        if (left[l] < right[r]) {
            array[i++] = left[l++];
        } else {
            array[i++] = right[r++];
        }
    }
    while (l < leftSize) {
        array[i++] = left[l++];
    }
    while (r < rightSize) {
        array[i++] = right[r++];
    }
}
// recursively calling part
void mergerSort(int *array , int length){
  // base condition
  if (length <= 1) return;
  int middle = length / 2;
  int *lefttArray = malloc(middle*sizeof(int));
  int *rightArray = malloc((length-middle)*sizeof(int));
  
  for (int i = 0; i < middle; i++) {
        leftArray[i] = array[i];
    }
  for (int i = middle; i < length; i++) {
      rightArray[i - middle] = array[i];
  }
  mergeSort(leftArray, middle);
  mergeSort(rightArray, length-middle);
  merge(leftArray, middle, rightArray, length - middle, array);
  // freeing memory
  free(leftArray);
  free(rightArray);  
}
// prtin the array
void printArray(int *array, int length){
  for (int i = 0 ; i < length ; i++){
    printf("%d",array[i]);
  }
}
int main(void){
  // sample array 
  int array[] = {9,8,7,6,5,4,3,2,1};
  int length = sizeof(array)/sizeof(array[0]);
  mergeSort(array, length);
  printArray(array, length);
  
  return 0;
}
